routers/api/v1/users.go

```diff
@@ -314,12 +314,12 @@ func (r *apiV1Router) Register(ctx *gin.Context) {
 	})
 }
 
-// POST: /api/v1/users/email?token={token}
-// Request:  token string
+// POST: /api/v1/users/email
 // Response: status int
 //           error string
+// Header:   Authorization -> token
 func (r *apiV1Router) VerifyEmail(ctx *gin.Context) {
-	token := ctx.Query("token")
+	token := ctx.GetHeader(authHeaderName)
```

Comment:
When a user verifies the email, shouldn't we extract the token from the parameter in the URL instead of using the auth token? Since when they click on the verification URL it won't set the Authorization header

Reply:
This is part of the API so it's not really meant to be called by just clicking a link in an email. It's a bit unsafe to store the auth token in the URL so I thought I'd use the same approach as the rest of the API and store the token in the header.

Storing the token in the URL is not avoidable for the front-end since we'll have to send the verification URL in an email but there's no need to have the same security vulnerability on the API.

Reply:
> This is part of the API so it's not really meant to be called by just clicking a link in an email. It's a bit unsafe to store the auth token in the URL so I thought I'd use the same approach as the rest of the API and store the token in the header.
> 
> Storing the token in the URL is not avoidable for the front-end since we'll have to send the verification URL in an email but there's no need to have the same security vulnerability on the API.

Couldn't we generate some unique uuid, and store it in the database under `password_reset_token` (or something like that) along with the time it was created. Then we can validate the reset token and make sure it was within the time limit. This means we don't have to worry about sending a JWT token in the URL and we also don't need to worry about the Authorisation header.

Maybe I misunderstood your answer, but I thought this way might be a bit easier. Let me know that you think 😊

Reply:
Well we'd need to store the token's uuid in the URL and that's not really any different than storing the JWT in the URL itself.
The problem I'm trying to avoid making the token available to anyone who can see the URL.

There is no real solution for this for the frontend implementation but on the frontend we can use a different endpoint (could just be a wrapper for this endpoint) for email verification that doesn't need to use headers but for the API there isn't any good reason why we'd need to store the token in the URL.

Also the expiration time can be stored in the token itself as well. Tbh, the implementation for token expiration is just going to be a couple of lines. I just haven't gotten around to implementing it yet

Reply:
True, makes sense 👍

Reply:
>Also the expiration time can be stored in the token itself as well. Tbh, the implementation for token expiration is just going to be a couple of lines. I just haven't gotten around to implementing it yet

Well okay, it depends on how safe we want the email tokens to be. If we're okay with just having them expire based on time, the implementation is very easy. If we want something more secure to make sure that a token can only be used once, we can include a checksum of the user in the token and then check if the checksum in the token matches the checksum of the user in the DB. This would involve quite a bit more effort and I'm not sure if it is actually necessary.

Reply:
I don't think we need to worry about one-time use tokens yet. Time based expiration is fine with me for now. It's something we can look into in the future.

Reply:
I was considering this for ordinary auth tokens because this would help with solving consistency issues if the user changes at any point. Because right now, if say `hs_application` has a cached version of a user and the user's auth level changes, `hs_application` would not be aware of the update and the token that `hs_application` has would also be valid. Storing the user's checksum in the token would allow us to make sure that a token is valid only if it matches the most recent version of the user.

This would not introduce any breaking changes to `hs_auth`'s consumers (as the checksum would be stored in the token and the validation logic is on `hs_auth`) so this can be left for the future.